src/aks-preview/azext_aks_preview/managed_cluster_decorator.py

```diff
@@ -4485,24 +4487,30 @@ def _update_dns_zone_resource_ids(self, mc: ManagedCluster, dns_zone_resource_id
 
         if mc.ingress_profile and mc.ingress_profile.web_app_routing and mc.ingress_profile.web_app_routing.enabled:
             if add_dns_zone:
-                if mc.ingress_profile.web_app_routing.dns_zone_resource_ids is None:
-                    mc.ingress_profile.web_app_routing.dns_zone_resource_ids = []
-                mc.ingress_profile.web_app_routing.dns_zone_resource_ids.extend(dns_zone_resource_ids)
-                if attach_zones:
-                    try:
-                        for dns_zone in dns_zone_resource_ids:
-                            if not add_role_assignment(
-                                self.cmd,
-                                "DNS Zone Contributor",
-                                mc.ingress_profile.web_app_routing.identity.object_id,
-                                False,
-                                scope=dns_zone
-                            ):
-                                logger.warning(
-                                    'Could not create a role assignment for App Routing. '
-                                    'Are you an Owner on this subscription?')
-                    except Exception as ex:
-                        raise CLIError('Error in granting dns zone permisions to managed identity.\n') from ex
+                mc.ingress_profile.web_app_routing.dns_zone_resource_ids = (
+                    mc.ingress_profile.web_app_routing.dns_zone_resource_ids or []
+                )
+                for dns_zone_id in dns_zone_resource_ids:
+                    if dns_zone_id not in mc.ingress_profile.web_app_routing.dns_zone_resource_ids:
+                        mc.ingress_profile.web_app_routing.dns_zone_resource_ids.append(dns_zone_id)
+                        if attach_zones:
+                            try:
+                                is_private_dns_zone = (
+                                    parse_resource_id(dns_zone_id).get("type").lower() == "privatednszones"
+                                )
+                                role = "Private DNS Zone Contributor" if is_private_dns_zone else "DNS Zone Contributor"
```

Comment:
It would be better to declare these two strings as constants and then reference them here

Reply:
added constants for DNS Zone Contributor roles.